File: src/data/repository/ExerciseRepository.ts
/**
 * Exercise Repository - Manages exercise library
 * Migrated from Android Kotlin ExerciseRepository to React Native TypeScript
 */

import * as ExerciseDao from '../local/daos/exerciseDao';
import { ExerciseEntity, ExerciseVideoEntity } from '../local/entities';

/**
 * Exercise Repository interface
 */
export interface IExerciseRepository {
  getAllExercises(): Promise<ExerciseEntity[]>;
  searchExercises(query: string): Promise<ExerciseEntity[]>;
  filterByMuscleGroup(muscleGroup: string): Promise<ExerciseEntity[]>;
  filterByEquipment(equipment: string): Promise<ExerciseEntity[]>;
  getFavorites(): Promise<ExerciseEntity[]>;
  toggleFavorite(id: string): Promise<void>;
  getExerciseById(id: string): Promise<ExerciseEntity | null>;
  getVideos(exerciseId: string): Promise<ExerciseVideoEntity[]>;
  importExercises(): Promise<void>;
  isExerciseLibraryEmpty(): Promise<boolean>;
}

/**
 * Exercise Repository implementation
 */
class ExerciseRepositoryImpl implements IExerciseRepository {
  constructor() {}

  /**
   * Get all exercises sorted by name
   */
  async getAllExercises(): Promise<ExerciseEntity[]> {
    try {
      return await ExerciseDao.getAllExercises();
    } catch (error) {
      console.error('[ExerciseRepository] Failed to get all exercises:', error);
      throw error;
    }
  }

  /**
   * Search exercises by name, description, or muscles
   */
  async searchExercises(query: string): Promise<ExerciseEntity[]> {
    try {
      if (!query || query.trim().length === 0) {
        return await this.getAllExercises();
      }
      return await ExerciseDao.searchExercises(query.trim());
    } catch (error) {
      console.error('[ExerciseRepository] Failed to search exercises:', error);
      throw error;
    }
  }

  /**
   * Filter exercises by muscle group
   */
  async filterByMuscleGroup(muscleGroup: string): Promise<ExerciseEntity[]> {
    try {
      if (!muscleGroup || muscleGroup.trim().length === 0) {
        return await this.getAllExercises();
      }
      return await ExerciseDao.getExercisesByMuscleGroup(muscleGroup);
    } catch (error) {
      console.error('[ExerciseRepository] Failed to filter by muscle group:', error);
      throw error;
    }
  }

  /**
   * Filter exercises by equipment
   */
  async filterByEquipment(equipment: string): Promise<ExerciseEntity[]> {
    try {
      if (!equipment || equipment.trim().length === 0) {
        return await this.getAllExercises();
      }
      return await ExerciseDao.getExercisesByEquipment(equipment);
    } catch (error) {
      console.error('[ExerciseRepository] Failed to filter by equipment:', error);
      throw error;
    }
  }

  /**
   * Get favorite exercises
   */
  async getFavorites(): Promise<ExerciseEntity[]> {
    try {
      return await ExerciseDao.getFavorites();
    } catch (error) {
      console.error('[ExerciseRepository] Failed to get favorites:', error);
      throw error;
    }
  }

  /**
   * Toggle favorite status for an exercise
   */
  async toggleFavorite(id: string): Promise<void> {
    try {
      const exercise = await ExerciseDao.getExerciseById(id);
      if (exercise) {
        await ExerciseDao.updateFavorite(id, !exercise.isFavorite);
        console.log(`[ExerciseRepository] Toggled favorite for exercise: ${id}`);
      }
    } catch (error) {
      console.error('[ExerciseRepository] Failed to toggle favorite:', error);
      throw error;
    }
  }

  /**
   * Get exercise by ID
   */
  async getExerciseById(id: string): Promise<ExerciseEntity | null> {
    try {
      return await ExerciseDao.getExerciseById(id);
    } catch (error) {
      console.error('[ExerciseRepository] Failed to get exercise by ID:', error);
      return null;
    }
  }

  /**
   * Get videos for an exercise
   */
  async getVideos(exerciseId: string): Promise<ExerciseVideoEntity[]> {
    try {
      return await ExerciseDao.getVideos(exerciseId);
    } catch (error) {
      console.error('[ExerciseRepository] Failed to get videos:', error);
      return [];
    }
  }

  /**
   * Import exercises from assets (if not already imported)
   * In React Native, this would typically import from a JSON file bundled with the app
   */
  async importExercises(): Promise<void> {
    try {
      // Check if exercises are already imported
      const count = await this.isExerciseLibraryEmpty();
      if (!count) {
        console.log('[ExerciseRepository] Exercises already imported');
        return;
      }

      console.log('[ExerciseRepository] Importing exercises from assets...');

      // Import exercises from bundled JSON
      // Note: In React Native, you would typically require a JSON file here
      // Example: const exercises = require('../../assets/exercises.json');
      // For now, we'll assume the import functionality is handled by a separate service
      // that calls ExerciseDao.insertExercise for each exercise

      console.log('[ExerciseRepository] Exercise import would be handled here');
      console.log('[ExerciseRepository] In production, load from bundled JSON and insert into database');
    } catch (error) {
      console.error('[ExerciseRepository] Failed to import exercises:', error);
      throw error;
    }
  }

  /**
   * Check if exercise library is empty
   */
  async isExerciseLibraryEmpty(): Promise<boolean> {
    try {
      const exercises = await this.getAllExercises();
      return exercises.length === 0;
    } catch (error) {
      console.error('[ExerciseRepository] Failed to check if library is empty:', error);
      return true;
    }
  }
}

// Export singleton instance
let exerciseRepositoryInstance: ExerciseRepositoryImpl | null = null;

export const getExerciseRepository = (): IExerciseRepository => {
  if (!exerciseRepositoryInstance) {
    exerciseRepositoryInstance = new ExerciseRepositoryImpl();
  }
  return exerciseRepositoryInstance;
};

export const resetExerciseRepository = (): void => {
  exerciseRepositoryInstance = null;
};

// Export types
export type { IExerciseRepository };
export { ExerciseRepositoryImpl };

File: deprecated/app/src/main/java/com/example/vitruvianredux/data/repository/ExerciseRepository.kt
package com.example.vitruvianredux.data.repository

import com.example.vitruvianredux.data.local.ExerciseDao
import com.example.vitruvianredux.data.local.ExerciseEntity
import com.example.vitruvianredux.data.local.ExerciseImporter
import com.example.vitruvianredux.data.local.ExerciseVideoEntity
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.firstOrNull
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for exercise library management
 */
interface ExerciseRepository {
    fun getAllExercises(): Flow<List<ExerciseEntity>>
    fun searchExercises(query: String): Flow<List<ExerciseEntity>>
    fun filterByMuscleGroup(muscleGroup: String): Flow<List<ExerciseEntity>>
    fun filterByEquipment(equipment: String): Flow<List<ExerciseEntity>>
    fun getFavorites(): Flow<List<ExerciseEntity>>
    suspend fun toggleFavorite(id: String)
    suspend fun getExerciseById(id: String): ExerciseEntity?
    suspend fun getVideos(exerciseId: String): List<ExerciseVideoEntity>
    suspend fun importExercises(): Result<Unit>
    suspend fun isExerciseLibraryEmpty(): Boolean
}

/**
 * Implementation of ExerciseRepository
 */
@Singleton
class ExerciseRepositoryImpl @Inject constructor(
    private val exerciseDao: ExerciseDao,
    private val exerciseImporter: ExerciseImporter
) : ExerciseRepository {
    
    /**
     * Get all exercises sorted by name
     */
    override fun getAllExercises(): Flow<List<ExerciseEntity>> {
        return exerciseDao.getAllExercises()
    }
    
    /**
     * Search exercises by name, description, or muscles
     */
    override fun searchExercises(query: String): Flow<List<ExerciseEntity>> {
        return if (query.isBlank()) {
            getAllExercises()
        } else {
            exerciseDao.searchExercises(query.trim())
        }
    }
    
    /**
     * Filter exercises by muscle group
     */
    override fun filterByMuscleGroup(muscleGroup: String): Flow<List<ExerciseEntity>> {
        return if (muscleGroup.isBlank()) {
            getAllExercises()
        } else {
            exerciseDao.getExercisesByMuscleGroup(muscleGroup)
        }
    }
    
    /**
     * Filter exercises by equipment
     */
    override fun filterByEquipment(equipment: String): Flow<List<ExerciseEntity>> {
        return if (equipment.isBlank()) {
            getAllExercises()
        } else {
            exerciseDao.getExercisesByEquipment(equipment)
        }
    }
    
    /**
     * Get favorite exercises
     */
    override fun getFavorites(): Flow<List<ExerciseEntity>> {
        return exerciseDao.getFavorites()
    }
    
    /**
     * Toggle favorite status for an exercise
     */
    override suspend fun toggleFavorite(id: String) {
        try {
            val exercise = exerciseDao.getExerciseById(id)
            exercise?.let {
                exerciseDao.updateFavorite(id, !it.isFavorite)
                Timber.d("Toggled favorite for exercise: $id")
            }
        } catch (e: Exception) {
            Timber.e(e, "Failed to toggle favorite")
        }
    }
    
    /**
     * Get exercise by ID
     */
    override suspend fun getExerciseById(id: String): ExerciseEntity? {
        return exerciseDao.getExerciseById(id)
    }
    
    /**
     * Get videos for an exercise
     */
    override suspend fun getVideos(exerciseId: String): List<ExerciseVideoEntity> {
        return exerciseDao.getVideos(exerciseId)
    }
    
    /**
     * Import exercises from assets (if not already imported)
     */
    override suspend fun importExercises(): Result<Unit> {
        return try {
            // Check if exercises are already imported
            val count = getAllExercises().firstOrNull()?.size ?: 0
            if (count == 0) {
                Timber.d("Importing exercises from assets...")
                val result = exerciseImporter.importExercises()
                if (result.isSuccess) {
                    Result.success(Unit)
                } else {
                    result.exceptionOrNull()?.let { Result.failure(it) } ?: Result.failure(Exception("Import failed"))
                }
            } else {
                Timber.d("Exercises already imported (count: $count)")
                Result.success(Unit)
            }
        } catch (e: Exception) {
            Timber.e(e, "Failed to import exercises")
            Result.failure(e)
        }
    }
    
    /**
     * Check if exercise library is empty
     */
    override suspend fun isExerciseLibraryEmpty(): Boolean {
        return getAllExercises().firstOrNull()?.isEmpty() ?: true
    }
}